Code changes to enable player thumbnail seekbar preview in NewPipe

litetex opened the pull request:
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [x] I have tested the API against [NewPipe](https://github.com/TeamNewPipe/NewPipe).
- [x] I agree to create a pull request for [NewPipe](https://github.com/TeamNewPipe/NewPipe) as soon as possible to make it compatible with the changed API.

This pr is required for fixing https://github.com/TeamNewPipe/NewPipe/issues/3207
This is the nested/required PR for https://github.com/TeamNewPipe/NewPipe/pull/6434 

Also fixes: #595

### Changes
* Exposed the thumbnail preview frames

TobiGr commented on extractor/src/main/java/org/schabi/newpipe/extractor/stream/Frameset.java:
```diff
@@ -1,18 +1,37 @@
-    private int framesPerPageY;
-
-    public Frameset(List<String> urls, int frameWidth, int frameHeight, int totalCount, int durationPerFrame, int framesPerPageX, int framesPerPageY) {
+/*
+ Note:
+ Implement Serializable here or
+ java.lang.RuntimeException: Parcelable encountered IOException writing serializable object
+  (name = org.schabi.newpipe.extractor.stream.StreamInfo)
+  ...
+ java.io.NotSerializableException: org.schabi.newpipe.extractor.stream.Frameset
+  ...
+ is thrown
+ */
```
That note is not needed
```suggestion

```

litetex replied:
I rebased the branch and removed it 😄

TobiGr reviewed (approved):
Looks good. Just one small thing.

Note at whoever this is merging: please squash

litetex commented:
Wait a moment, when a very short video is played (e.g. https://www.youtube.com/watch?v=jhFDyDgMVUI) then the extractor fails:
```
-------------------------------------
org.schabi.newpipe.extractor.exceptions.ExtractionException: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String[] java.lang.String.split(java.lang.String)' on a null object reference
	at org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeStreamExtractor.getFrames(YoutubeStreamExtractor.java:1099)
	at org.schabi.newpipe.extractor.stream.StreamInfo.extractOptionalData(StreamInfo.java:339)
	at org.schabi.newpipe.extractor.stream.StreamInfo.getInfo(StreamInfo.java:73)
	at org.schabi.newpipe.extractor.stream.StreamInfo.getInfo(StreamInfo.java:64)
	at org.schabi.newpipe.util.ExtractorHelper.lambda$getStreamInfo$3(ExtractorHelper.java:115)
	at org.schabi.newpipe.util.-$$Lambda$ExtractorHelper$YTHJjScxCJNO1LTCqs3IKy35iyY.call(Unknown Source:4)
	at io.reactivex.rxjava3.internal.operators.single.SingleFromCallable.subscribeActual(SingleFromCallable.java:43)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4813)
	at io.reactivex.rxjava3.internal.operators.single.SingleDoOnSuccess.subscribeActual(SingleDoOnSuccess.java:35)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4813)
	at io.reactivex.rxjava3.internal.operators.maybe.MaybeFromSingle.subscribeActual(MaybeFromSingle.java:41)
	at io.reactivex.rxjava3.core.Maybe.subscribe(Maybe.java:5330)
	at io.reactivex.rxjava3.internal.operators.maybe.MaybeConcatArray$ConcatMaybeObserver.drain(MaybeConcatArray.java:153)
	at io.reactivex.rxjava3.internal.operators.maybe.MaybeConcatArray$ConcatMaybeObserver.request(MaybeConcatArray.java:78)
	at io.reactivex.rxjava3.internal.operators.flowable.FlowableElementAtMaybe$ElementAtSubscriber.onSubscribe(FlowableElementAtMaybe.java:66)
	at io.reactivex.rxjava3.internal.operators.maybe.MaybeConcatArray.subscribeActual(MaybeConcatArray.java:42)
	at io.reactivex.rxjava3.core.Flowable.subscribe(Flowable.java:15753)
	at io.reactivex.rxjava3.internal.operators.flowable.FlowableElementAtMaybe.subscribeActual(FlowableElementAtMaybe.java:36)
	at io.reactivex.rxjava3.core.Maybe.subscribe(Maybe.java:5330)
	at io.reactivex.rxjava3.internal.operators.maybe.MaybeToSingle.subscribeActual(MaybeToSingle.java:46)
	at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4813)
	at io.reactivex.rxjava3.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)
	at io.reactivex.rxjava3.core.Scheduler$DisposeTask.run(Scheduler.java:614)
	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:65)
	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:56)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
	at java.lang.Thread.run(Thread.java:923)
Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String[] java.lang.String.split(java.lang.String)' on a null object reference
	at org.schabi.newpipe.extractor.services.youtube.extractors.YoutubeStreamExtractor.getFrames(YoutubeStreamExtractor.java:1061)
	... 29 more
-------------------------------------
```

TobiGr commented:
Something simple like TobiGr/NewPipeExtractor@e831c0e5613648ecf85e0c1457b570241684d7d6 should solve that

Stypox commented:
@litetex were you able to find out the problem?

litetex commented:
@Stypox 
> @litetex were you able to find out the problem?

Yes, as stated above, it was already fixed with fb7228a 😄

TobiGr commented:
Please replace `new ArrayList<>()` with `Collection.emptyList()`.

litetex commented on extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/extractors/YoutubeStreamExtractor.java:
```diff
@@ -1058,7 +1058,16 @@ private static String getVideoInfoUrl(final String id, final String sts) {
+
+            final String storyboardsRendererSpec = storyboardsRenderer.getString("spec");
+            if (storyboardsRendererSpec == null) {
+                return new ArrayList<>();
```
@litetex Self note: Replace with ``Collections.emptyList``

litetex commented:
@TobiGr 
> Please replace `new ArrayList<>()` with `Collection.emptyList()`.

Done 😄

Stypox commented on extractor/src/main/java/org/schabi/newpipe/extractor/stream/StreamInfo.java:
```diff
@@ -386,6 +392,11 @@ private static StreamInfo extractOptionalData(StreamInfo streamInfo, StreamExtra
     private List<StreamSegment> streamSegments = new ArrayList<>();
     private List<MetaInfo> metaInfo = new ArrayList<>();
 
+    /**
+     * Preview frames, e.g. for the storyboard / seekbar thumbnail preview
+     */
+    private List<Frameset> previewFrames = new ArrayList<>();
```
I think these should also all be `emptyList()`, since they are never modified but only get/set.

litetex replied:
Why are all other fields (above) then also ``new ArrayList<>()``? 
Shouldn't these be changed to?

Stypox replied:
Yes, they should

litetex commented:
Rebased the branch again and applied @Stypox changes.

Any change of getting this merged soon?
Getting tired of rebasing the branch every week 😢

Stypox commented:
> Any change of getting this merged soon?

After 0.21.6 is released I'd merge this right away @TobiGr
